File: 51-n-queens/51-n-queens.java
class Solution {
    public List<List<String>> solveNQueens(int n) {
        char[][] chess = new char[n][n];
        for(char[] row : chess) Arrays.fill(row, '.');
        
        List<List<String>> ans = new ArrayList<>();
        HashSet<Integer> plusDiagonal = new HashSet<>();
        HashSet<Integer> minusDiagonal = new HashSet<>();
        HashSet<Integer> columns = new HashSet<>();
        putQueen(0, n, chess, ans, plusDiagonal, minusDiagonal, columns);
        return ans;
    }
    
    public void putQueen(int row, int n, char[][] chess, List<List<String>> ans, HashSet<Integer> plusDiagonal, HashSet<Integer> minusDiagonal, HashSet<Integer> columns){
        if(row == n){
            List<String> possible = new ArrayList<>();
            for(char[] r : chess){
                possible.add(new String(r));
            }
            ans.add(possible);
            return;
        }
        
        for(int col = 0; col < n; col++){
            if(!plusDiagonal.contains(row + col) && !minusDiagonal.contains(row - col) && !columns.contains(col)){
                plusDiagonal.add(row + col);
                minusDiagonal.add(row - col);
                columns.add(col);
                
                chess[row][col] = 'Q';                 
                putQueen(row + 1, n, chess, ans, plusDiagonal, minusDiagonal, columns);
                chess[row][col] = '.';
                
                plusDiagonal.remove(row + col);
                minusDiagonal.remove(row - col);
                columns.remove(col);
            }
        }
    }
}

File: 51-n-queens/51-n-queens.py
class Solution:
    def solveNQueens(self, n: int) -> List[List[str]]:
        blockedColumn = set()
        minusDiagonal = set()
        plusDiagonal = set()
        
        chessboard = [['.'] * n for col in range(n)]
        ans = []
        
        self.canPutQueen(0, n, chessboard, blockedColumn, minusDiagonal, plusDiagonal, ans)
        
        return ans
    
    def canPutQueen(self, current_row, n, chessboard, blockedColumn, minusDiagonal, plusDiagonal, ans):
        if current_row == n:
            pos = []
            for i in range(len(chessboard)):
                pos.append(''.join(chessboard[i]))
            ans.append(pos)
            return
        
        for col in range(n):
            if col in blockedColumn or (current_row + col) in plusDiagonal or (current_row - col) in minusDiagonal:
                continue
            
            blockedColumn.add(col)
            plusDiagonal.add(current_row + col)
            minusDiagonal.add(current_row - col)
            
            chessboard[current_row][col] = 'Q';
            
            self.canPutQueen(current_row + 1, n, chessboard,  blockedColumn, minusDiagonal, plusDiagonal, ans)
            
            chessboard[current_row][col] = '.';
            blockedColumn.remove(col)
            plusDiagonal.remove(current_row + col)
            minusDiagonal.remove(current_row - col)
        
        